### PR TITLE
nixos/keter: 2.0 -> 2.1

### PR DESCRIPTION
**Pull request by jappeace**
keter 2.1 now can log to stderr instead of file rotation.
Which is faster and more reliable.
These changes support that.

Announcement:
https://discourse.haskell.org/t/keter-2-1-0-released/6134

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant **TODO**, this should pass
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by jappeace** [comment]
dear fren @ckiee , if you have time please consider reviewing this once more.

**Review by ckiee — CHANGES_REQUESTED** [review]

File: nixos/modules/services/web-servers/keter/default.nix [hunk 1]
```diff
@@ -34,17 +34,21 @@ Keep an old app running and swap the ports when the new one is booted.
           host = "*4";
```
[the remark this answers is not in the document]
Comment (jappeace) [reply nixos/modules/services/web-servers/keter/default.nix:34]: keter 2.1 will no longer log properly with the current nixpkgs module. These change will make the logs behave properly by relying on the newly created stderr features (which just dropped the NIH solution and uses well maintained libraries).

File: nixos/modules/services/web-servers/keter/default.nix [hunk 2]
```diff
@@ -34,17 +34,21 @@ Keep an old app running and swap the ports when the new one is booted.
...
           port = 6981;
         }];
+        rotate-logs= false;
       };
       # You want that ip-from-header in the nginx setup case
       # so it's not set to 127.0.0.1.
       # using a port above 1024 allows you to avoid needing CAP_NET_BIND_SERVICE
+      # rotate-logs = false; emits keter logs and it's applications to stderr
+      # which allows journald to capture them.
```
[the remark this answers is not in the document]
Comment (jappeace) [reply nixos/modules/services/web-servers/keter/default.nix:43]: L39 to 41 seemed to explain the defaultText options in a similar way, I followed this pattern.

File: nixos/tests/keter.nix [hunk 3]
```diff
@@ -18,6 +18,7 @@ in
           host = "*4";
           inherit port;
         }];
+        rotate-logs= false;
```
[the remark this answers is not in the document]
Comment (jappeace) [reply nixos/tests/keter.nix:21]: yes I ran nixpkgs-fmt

File: nixos/modules/services/web-servers/keter/default.nix (hunk 2 above)
[the remark this answers is not in the document]
Comment (ckiee) [reply nixos/modules/services/web-servers/keter/default.nix:43]: It's.. your own module? And therefore your own pattern. Please consult other modules.

File: nixos/modules/services/web-servers/keter/default.nix (hunk 1 above)
[the remark this answers is not in the document]
Comment (ckiee) [reply nixos/modules/services/web-servers/keter/default.nix:34]: Ah, and keter comes from an auto-generated & updated `haskellPackages` (/hackage). I see.

File: nixos/modules/services/web-servers/keter/default.nix (hunk 2 above)
[the remark this answers is not in the document]
Comment (jappeace) [reply nixos/modules/services/web-servers/keter/default.nix]: yeah, I've to admit, I don't remember what past me did or why, anyway, I moved it to L37.
there is no option description for this particular option since it just emits the global config as a yaml file and let's keter deal with parsing the options.

File: nixos/modules/services/web-servers/keter/default.nix (hunk 1 above)
[the remark this answers is not in the document]
Comment (jappeace) [reply nixos/modules/services/web-servers/keter/default.nix:34]: that's correct.

File: nixos/tests/keter.nix (hunk 3 above)
[the remark this answers is not in the document]
Comment (jappeace) [reply nixos/tests/keter.nix]: yeah, I fucked up, I didn't realize this was a seperate file, turns out it never had been formatted, so the diff is a bit larger now.

File: nixos/modules/services/web-servers/keter/default.nix [hunk 4]
```diff
@@ -34,6 +34,9 @@ Keep an old app running and swap the ports when the new one is booted.
           host = "*4";
           port = 6981;
         }];
+        # rotate-logs = false; emits keter logs and it's applications to stderr
+        # which allows journald to capture them.
+        rotate-logs = false;
       };
```
Comment (ckiee) [inline nixos/modules/services/web-servers/keter/default.nix:40]: I'm not sure if `default` for this option is going to be recursive. Please observe [this pattern](https://github.com/ckiee/nixpkgs/blob/3e272a4f63a46813bf95e3fca23684dc8049d9a3/nixos/modules/services/web-servers/garage.nix#L33-L59) where we can document individual options of a `settings`-esque configuration.

File: nixos/tests/keter.nix [hunk 5]
```diff
@@ -1,42 +1,43 @@
 import ./make-test-python.nix ({ pkgs, ... }:
-let
-  port = 81;
-in
-{
-  name = "keter";
-  meta = with pkgs.lib.maintainers; {
-    maintainers = [ jappie ];
-  };
+  let
```
Comment (ckiee) [inline nixos/tests/keter.nix:2]: Please keep the reformatting in another commit next time.

File: nixos/modules/services/web-servers/keter/default.nix (hunk 4 above)
Comment (ckiee) [reply nixos/modules/services/web-servers/keter/default.nix:40]: To be clearer, what I mean is I don't know what if `rotate-logs`'s default would stay as `false` when you change something else under `globalKeterConfig`. Like if you set:
```nix
services.keter.globalKeterConfig.some-other-option = "blep";
```
Will that keep `rotate-logs`? I'm rusty.

File: nixos/modules/services/web-servers/keter/default.nix [hunk 6]
```diff
@@ -34,6 +34,9 @@ Keep an old app running and swap the ports when the new one is booted.
           host = "*4";
```
Comment (ckiee) [inline nixos/modules/services/web-servers/keter/default.nix:34]: L29: `globalKeterConfig` -> `settings` perhaps? And why is it `keterPackage` and `keterRoot` also? This is `services.keter.*`; keter is implied.

Comment (ckiee) [inline nixos/modules/services/web-servers/keter/default.nix:34]: L30: `lib.types.attrs` is wrong, it's `(pkgs.formats.yaml {}).type` or something like that.

**Review by ckiee — CHANGES_REQUESTED** [review]
I wrote a few more things but I'm not going to be available to see this PR through or give it a full thorough review (surprised by how many things I missed [last time around](https://github.com/NixOS/nixpkgs/pull/162484)!). There's [a Discourse thread](https://discourse.nixos.org/t/prs-ready-for-review/3032) you can post your PR on.

Good luck!

File: nixos/modules/services/web-servers/keter/default.nix (hunk 6 above)
Comment (ckiee) [reply nixos/modules/services/web-servers/keter/default.nix:34]: You'll need to use `mkRenamedOptionModule`.

**Comment by jappeace** [comment]
@ckiee thanks for your effort! I'll find someone else, don't you worry :wink: 
feel free to ignore my replies, I'm just replying so other reviewers can see what I did.

File: nixos/modules/services/web-servers/keter/default.nix (hunk 6 above)
Comment (jappeace) [reply nixos/modules/services/web-servers/keter/default.nix]: yes I got this thanks!

Comment (jappeace) [reply nixos/modules/services/web-servers/keter/default.nix]: got it, thanks!

File: nixos/modules/services/web-servers/keter/default.nix (hunk 4 above)
Comment (jappeace) [reply nixos/modules/services/web-servers/keter/default.nix:73]: I copied the pattern as you suggested.
This is much nicer in usage.

File: nixos/modules/services/web-servers/keter/default.nix [hunk 7]
```diff
@@ -122,7 +146,7 @@ Keep an old app running and swap the ports when the new one is booted.
         script = ''
           set -xe
           mkdir -p ${incoming}
-          { tail -F ${cfg.keterRoot}/log/keter/current.log -n 0 & ${cfg.keterPackage}/bin/keter ${globalKeterConfigFile}; }
+          ${cfg.package}/bin/keter ${globalKeterConfigFile};
```
Comment (SuperSandro2000) [inline nixos/modules/services/web-servers/keter/default.nix:149]: We could use lib.getExe here

File: nixos/modules/services/web-servers/keter/default.nix [hunk 8]
```diff
@@ -1,53 +1,77 @@
...
+                  description =  lib.mdDoc "port";
+                };
+              };
+            });
+            description = lib.mdDoc "You want that ip-from-header in the nginx setup case. It allows nginx setting the original ip address rather then it being localhost (due to reverse proxying)";
+          };
+          rotate-logs = lib.mkOption {
+            default = false;
+            type = lib.types.bool;
+            description = lib.mdDoc "rotate-logs = false; emits keter logs and it's applications to stderr. which allows journald to capture them. Set to true to let keter put the logs in files (useful on non systemd systems)";
```
Comment (SuperSandro2000) [inline nixos/modules/services/web-servers/keter/default.nix:70]: This should be multi line. Also the last sentence is creatine more questions for me than solving

File: nixos/modules/services/web-servers/keter/default.nix [hunk 9]
```diff
@@ -1,53 +1,77 @@
...
-          host = "*4";
-          port = 6981;
-        }];
+      type = lib.types.submodule {
+        freeformType = yaml.type;
+        options = {
+          ip-from-header = lib.mkOption {
+            default = true;
+            type = lib.types.bool;
+            description = lib.mdDoc "You want that ip-from-header in the nginx setup case. It allows nginx setting the original ip address rather then it being localhost (due to reverse proxying)";
```
Comment (SuperSandro2000) [inline nixos/modules/services/web-servers/keter/default.nix:49]: Why is it then even an option?

File: nixos/modules/services/web-servers/keter/default.nix [hunk 10]
```diff
@@ -1,53 +1,77 @@
...
+  yaml = pkgs.formats.yaml {};
 in
 {
   meta = {
     maintainers = with lib.maintainers; [ jappie ];
   };
 
+  imports = [
+    (lib.mkRenamedOptionModule
+      [ "services" "keter" "keterRoot"]
```
Comment (SuperSandro2000) [inline nixos/modules/services/web-servers/keter/default.nix:13]: We normally put this on one line

File: nixos/modules/services/web-servers/keter/default.nix (hunk 9 above)
Comment (jappeace) [reply nixos/modules/services/web-servers/keter/default.nix:49]: keter doesn't neccisarly need to be put behind nginx, it apparantly also has https management.
but since the client (=supercede) already had nginx setup, 
I personally never configured it to be used like that.

File: nixos/modules/services/web-servers/keter/default.nix (hunk 10 above)
Comment (jappeace) [reply nixos/modules/services/web-servers/keter/default.nix]: I put these on one line

File: nixos/modules/services/web-servers/keter/default.nix (hunk 8 above)
Comment (jappeace) [reply nixos/modules/services/web-servers/keter/default.nix]: I made these multi line

File: nixos/modules/services/web-servers/keter/default.nix (hunk 7 above)
Comment (jappeace) [reply nixos/modules/services/web-servers/keter/default.nix]: it's using getExe now

**Comment by jappeace** [comment]
@SuperSandro2000 I've squashed all changes into nixos/keter, except for the run nixpkgs-fmt since I've been explicetly told to keep format commits separate.
I also prefixed that commit with `nixos/keter:`

If your comment tries to address some other issue, please name your specific concern, since it's hard to figure out what you mean specifically by citing a 1500 word count document.

File: nixos/modules/services/web-servers/keter/default.nix [hunk 11]
```diff
@@ -1,53 +1,82 @@
...
+          listeners = lib.mkOption {
+            default = [{ host = "*"; port = 6981; }];
+            type = lib.types.listOf (lib.types.submodule {
+              options = {
+                host = lib.mkOption {
+                  type = lib.types.str;
+                  description = lib.mdDoc "host";
+                };
+                port = lib.mkOption {
+                  type = lib.types.int;
```
Comment (SuperSandro2000) [inline nixos/modules/services/web-servers/keter/default.nix:54]:
```suggestion
                  type = lib.types.port;
```

**Review by SuperSandro2000 — APPROVED** [review]
